### ISSUE TITLE
Using . (dot) in context_before eats up that character

### ISSUE THREAD

**Issue by joanise** [roedoejet/g2p#31]
**TLDR**: if a rule has `context_before = "."`, the character before `in` is incorrectly deleted.

This is clearly related to issue #15 but apparently I found another situation that trips that patch.

**Long story**:

In French, I want a rule that says delete word-final "s" as long as it's not also word initial, i.e., as long as "s" is not the whole word.

In branch `dev.fra`, commit `0233c15`, I have this rule in `fra/fra_to_ipa.csv` that attempts to accomplish that:
```
s,,.,\s|$
```
it replaces my rule `s,,,\s|$` in the same file in branch `master` but attempts to avoid having it applied to "s".

However, when this rule is applied, it erases both the "s" and the character that precedes it:
```
$ g2p convert écoutons fra fra-ipa --debugger
[...]
    [   {   'end': 9,
            'input': 'écoutons',
            'output': 'écouto',
            'rule': {   'context_after': '\\s|$',
                        'context_before': '.',
                        'in': 's',
                        'out': ''},
            'start': 7},
[...]
```
As shown above, the output of applying my rule is "écouto" instead of "écouton" as intended.

**Comment by roedoejet** [comment]
So it looks like periods weren't being included when creating the lookbehinds, so the regex being created was `re.compile('.s(?=\\s|$)', re.IGNORECASE)` and not `re.compile('((?<=.))s(?=\\s|$)', re.IGNORECASE)` as it should be.

I fixed that here: 4857c01 - perhaps there are other characters to add though? I'm a little brain-tired right now, but maybe we can discuss this tomorrow! I won't close this issue quite yet though.

Also, btw, is there any reason you're working on your fork? Could you switch to just working on master[dev.fra]?

**Comment by joanise** [comment]
Great, thanks. I'll test it.

I'm working on my fork when I create throw-away commits with garbage in them. I like to commit often, even crap, but I don't like having all that survive in the history. I don't mind having a bunch of commits if they're all meaningful, each adding at least something meaningful, but I don't want the kind of garbage that's currently in my fork's copy of `dev.fra`. I've been working with a private dev repo for years like that, and my GitHub fork is serving that purpose right now. I would not normally ask you to look at commits there, but this case was an exception since you wanted to see my broken LettersAccents abbreviation. Once it's referred to in a ticket, though, it's broken code that's worth saving in the history, so I guess it would have been appropriate to move that example code to origin/dev.fra.

Anyway, long story short, yes, I do work on the main repo's `master` and `dev.fra` by default.

**Comment by joanise** [comment]
Just pushed a commit with the rule showing the bug on the main `dev.fra`, as well as one with the broken abbreviations I reported elsewhere. Also changed the bug description to refer to that commit instead of my fork's.
Aidan, you're right, anything worth discussing and sharing belongs on the main repo, not my fork.

**Comment by joanise** [comment]
Back to the issue itself, are `\s`, `\S`, `\W`, `\w` and `\d` supported? I would have a need for `\S`, I think it would be safer than `.` for my non-initial context before rule.

**Comment by roedoejet** [comment]
Re: using the main repo, the above sounds good, I think it's totally fine to have a garbage repo, that makes sense. But yea, thanks also for committing the bug on `dev.fra` here too. I just found the bug. The regex *should* capture any sequence of backslash and then a character and put it in the lookbehind. Where it's getting tripped up though, is because `case_sensitive=false` is global for the mapping. So, what do you think, should we allow `case_sensitive to be local to a specific rule and then have `case_sensitive=true` for that rule, or should we try and avoid turning any sequence of backslash and a character into lowercase? I kind of feel like we should just have `case_sensitive=true` for that rule, although it does complicate the rule making process. What do you think?

**Comment by joanise** [comment]
You're welcome, for moving the commit to the main repo. It just makes sense.

As for `case_sensitive` by rule, I don't like it. `\S` is not a letter, it's an RE symbol. And I say match "s" after "\S", I still want "s" to be case insensitive. I believe it is accurate to claim that you only need to protect one letter after each backslash, I'm not aware of any other use of letters in REs that we could want to protect. Actually, why don't you use `re.match`'s own case insensitive option instead of lower-casing the expression?

**Comment by dhdaines** [comment]
shouldn't the dot be backslash-escaped in the regexes? otherwise it will just match any character

**Comment by dhdaines** [comment]
oh, wait, sorry! that's actually what you want :)

**Comment by joanise** [comment]
@roedoejet The dot now works correctly, as expected, but not in all contexts. Sorry to report that this saga is not over yet...

In `readalongs align`, my "s" now goes to "S" in eng-arpabet, as desired. And if I run `g2p convert s fra fra-ipa` and `g2p convert tests fra fra-ipa`, I get the expected output (`s` and `tʌst`).

Now, if I run `g2p convert "tests s" fra fra-ipa` the "s" vanishes in the second one, and the output is `tʌst` instead of `tʌst s`.

The problem is that my rule says `context_before = "."`, but what I'd really want is `context_before = "\S"`, which is not currently supported.

**Comment by roedoejet** [comment]
OK, makes sense. I'll close this and I've opened a separate issue related to handling case insensitive rules. Thanks @joanise !
